arangod/Scheduler/ThreadPoolScheduler.cpp

```diff
@@ -0,0 +1,138 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+#include "ThreadPoolScheduler.h"
+
+#include <velocypack/Builder.h>
+#include <numeric>
+
+#include "Metrics/Gauge.h"
+#include "Metrics/Counter.h"
+
+using namespace arangodb;
+
+void ThreadPoolScheduler::toVelocyPack(velocypack::Builder& b) const {
+  b.add("scheduler-threads", VPackValue(0));  // numWorkers
+  b.add("blocked", VPackValue(0));            // obsolete
+  b.add("queued", VPackValue(0));             // scheduler queue length
+  b.add("in-progress",
+        VPackValue(0));                 // number of working (non-idle) threads
+  b.add("direct-exec", VPackValue(0));  // obsolete
+}
+Scheduler::QueueStatistics ThreadPoolScheduler::queueStatistics() const {
+  return QueueStatistics();
+}
+void ThreadPoolScheduler::trackCreateHandlerTask() noexcept {
+  ++_metrics->_metricsHandlerTasksCreated;
+}
+void ThreadPoolScheduler::trackBeginOngoingLowPriorityTask() noexcept {
+  _metrics->_ongoingLowPriorityGauge += 1;
+}
+void ThreadPoolScheduler::trackEndOngoingLowPriorityTask() noexcept {
+  _metrics->_ongoingLowPriorityGauge -= 1;
+}
+void ThreadPoolScheduler::trackQueueTimeViolation() {
+  ++_metrics->_metricsQueueTimeViolations;
+}
+void ThreadPoolScheduler::trackQueueItemSize(std::int64_t x) noexcept {
+  _metrics->_schedulerQueueMemory += x;
+}
+
+uint64_t ThreadPoolScheduler::getLastLowPriorityDequeueTime() const noexcept {
+  return _lastLowPriorityDequeueTime.load(std::memory_order::relaxed);
+}
+
+void ThreadPoolScheduler::setLastLowPriorityDequeueTime(
+    uint64_t time) noexcept {
+  _lastLowPriorityDequeueTime.store(time, std::memory_order::relaxed);
+}
+
+std::pair<uint64_t, uint64_t>
+ThreadPoolScheduler::getNumberLowPrioOngoingAndQueued() const {
+  auto dequeued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.dequeued.load();
+  auto queued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.queued.load();
+  auto remaining = dequeued > queued ? 0 : queued - dequeued;
+  return std::make_pair(
+      _metrics->_ongoingLowPriorityGauge.load(std::memory_order_relaxed),
+      remaining);
+}
+
+// queue fill grad tracking is currently not implemented for this scheduler
+double ThreadPoolScheduler::approximateQueueFillGrade() const { return 0; }
+double ThreadPoolScheduler::unavailabilityQueueFillGrade() const { return 0; }
```

Comment:
Note that the queue fill grade may be checked by a load balancer periodically to determine if the instance is still in good shape.
Consider this calling code, which determines the instance's availability for further requests, from a load balancer's perspective:
```
        // if the scheduler's queue is more than x% full, render
        // the server unavailable
        double unavailabilityFillGrade =
            scheduler->unavailabilityQueueFillGrade();
        if (unavailabilityFillGrade > 0.0) {
          double fillGrade = scheduler->approximateQueueFillGrade();
          if (fillGrade >= unavailabilityFillGrade) {
            // oops, queue is relatively full
            available = false;
          }   
        }   
```
While this code works if the unavailability fill grade is 0, it will lead to more requests being sent to the instance now even if the queue is relatively full.
If this is an intentional change, fine. If not, is it much work to implement the fill grade methods in the new scheduler, or is there a reason why we shouldn't do this?

Reply:
These are functions defined in the Scheduler interface, but currently only make sense for the SupervisedScheduler (leaky abstraction). We will have to discuss if/how to add this to the new scheduler.